refactor - add belongsTo relationship for foreman_task

isratrade opened the pull request.

jwmatthews commented:
@isratrade I tested with a scratch build from @jmontleon (built at 11am EST) ruby193-rubygem-fusor_ui-0.0.1-94.el7.git.15.1fa5c25None.noarch.rpm 

Still seeing I'm stuck at 1C, can't proceed to 1D.

No errors noted in javascript console

From /var/log/foreman/production.log 
2015-12-21 16:33:49 [I] Processing by Fusor::Api::V21::DeploymentsController#create as JSON
2015-12-21 16:33:49 [I]   Parameters: {"deployment"=>{"name"=>"deploy_jwm2", "description"=>nil, "deploy_rhev"=>true, "deploy_cfme"=>true, "deploy_openstack"=>false, "is_disco
nnected"=>false, "rhev_is_self_hosted"=>false, "rhev_engine_admin_password"=>"[FILTERED]", "rhev_database_name"=>"Default", "rhev_cluster_name"=>"Default", "rhev_storage_name"
=>"my_storage", "rhev_storage_type"=>"NFS", "rhev_storage_address"=>nil, "rhev_cpu_type"=>nil, "rhev_share_path"=>nil, "cfme_install_loc"=>nil, "rhev_root_password"=>"[FILTERE
D]", "cfme_root_password"=>"[FILTERED]", "cfme_admin_password"=>"[FILTERED]", "foreman_task_uuid"=>nil, "upstream_consumer_uuid"=>nil, "upstream_consumer_name"=>nil, "rhev_exp
ort_domain_name"=>"my_export", "rhev_export_domain_address"=>nil, "rhev_export_domain_path"=>nil, "rhev_local_storage_path"=>nil, "rhev_gluster_node_name"=>nil, "rhev_gluster_
node_address"=>nil, "rhev_gluster_ssh_port"=>nil, "rhev_gluster_root_password"=>"[FILTERED]", "host_naming_scheme"=>nil, "custom_preprend_name"=>nil, "enable_access_insights"=
>false, "cfme_address"=>nil, "cfme_hostname"=>nil, "openstack_undercloud_password"=>"[FILTERED]", "openstack_undercloud_ip_addr"=>nil, "openstack_undercloud_user"=>nil, "opens
tack_undercloud_user_password"=>"[FILTERED]", "openstack_undercloud_hostname"=>nil, "openstack_overcloud_hostname"=>nil, "openstack_overcloud_address"=>nil, "openstack_overclo
ud_password"=>"[FILTERED]", "openstack_overcloud_private_net"=>nil, "openstack_overcloud_float_net"=>nil, "openstack_overcloud_float_gateway"=>nil, "cdn_url"=>nil, "manifest_f
ile"=>nil, "created_at"=>nil, "updated_at"=>nil, "rhev_engine_host_id"=>nil, "organization_id"=>"1", "lifecycle_environment_id"=>nil, "discovered_host_id"=>nil, "foreman_task_
id"=>nil}, "api_version"=>"v21"}
2015-12-21 16:33:49 [E] ActiveRecord::UnknownAttributeError: unknown attribute: foreman_task_id
/opt/rh/ruby193/root/usr/share/gems/gems/activerecord-3.2.8/lib/active_record/attribute_assignment.rb:88:in `block in assign_attributes'
/opt/rh/ruby193/root/usr/share/gems/gems/activerecord-3.2.8/lib/active_record/attribute_assignment.rb:78:in `each'
/opt/rh/ruby193/root/usr/share/gems/gems/activerecord-3.2.8/lib/active_record/attribute_assignment.rb:78:in `assign_attributes'
/opt/rh/ruby193/root/usr/share/gems/gems/activerecord-3.2.8/lib/active_record/base.rb:498:in `initialize'
/opt/rh/ruby193/root/usr/share/gems/gems/fusor_server-0.0.1/app/controllers/fusor/api/v21/deployments_controller.rb:31:in `new'
/opt/rh/ruby193/root/usr/share/gems/gems/fusor_server-0.0.1/app/controllers/fusor/api/v21/deployments_controller.rb:31:in `create'

jwmatthews commented:
@isratrade ignore my previous comment, looks like was related to browser caching the UI code I'm running.  I just kicked off a deployment, so no longer blocked on step 1C->1D.  

Will update later when deployment has completed to let you know about progress/next button.

jmontleon commented:
I had a fully successful RHEV/CFME/OSP deployment with the latest commits added as well.

isratrade commented:
Good news. Thanks for testing.
On Dec 21, 2015 8:23 PM, "jmontleon" <notifications@github.com> wrote:

> I had a fully successful RHEV/CFME/OSP deployment with the latest commits
> added as well.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fusor/fusor/pull/566#issuecomment-166381856>.
>

jwmatthews commented:
@isratrade my RHEV + CFME was successful as well.  Progress bars and next button look good.

isratrade commented:
@jwmatthews, thanks, I removed the "DON'T MERGE" from the PR title

isratrade commented on fusor-ember-cli/app/routes/deployment.js:
```diff
@@ -69,8 +69,7 @@ export default Ember.Route.extend(DeploymentRouteMixin, {
               resolve(response);
-              var uuid = response.id;
-              deployment.set('foreman_task_uuid', uuid);
+              deployment.set('foreman_task', response);
```
@jwmatthews, I added another commit that sets the belongsTo relationship `foreman_task` rather than the attribute that is no longer `foreman_task_uuid`. Does this work?

jwmatthews commented:
@isratrade I think we broke something with the last commit.

I don't see progress bars updating.

Additionally the issue is still present where if we navigate away from a running deployment we go back to step 5A, step 5B is disabled.

..but bigger problem now is that progress bars are not updating for me.  Deployment is progressing, but I don't see it reflected in progress bars.

isratrade commented:
@jwmatthews, ok, thanks. will look into this more. glad we didn't merge it.

isratrade commented:
@jwmatthews, a big help before your break if if you can send the the JSON response after successfully clicking the "Deploy" button

This is in the controller 
```ruby
      unless manifest_task["result"] == "error"
        task = async_task(::Actions::Fusor::Deploy,
                          @deployment,
                          params[:skip_content])
      end
      respond_for_async :resource => task
```

Is the JSON response the same as the GET response for a foreman-task?

jwmatthews commented:
@isratrade please see below for JSON response

I think below is from hitting the deploy button, took ~30 second for the request to finish a response to be generated.

Request URL:https://sat61fusor.example.com/fusor/api/v21/deployments/2/deploy

Response:

{"id":"b11fe7d5-6233-4938-b414-c815fb62242b","label":"Actions::Fusor::Deploy","pending":true,"username":"admin","started_at":"2015-12-24T02:10:52Z","ended_at":null,"state":"running","result":"pending","progress":0.0,"input":{"deployment_name":"deploy_12_23_a","locale":"en"},"output":{},"humanized":{"action":"Deploy deploy_12_23_a","input":[],"output":"","errors":[]},"cli_example":null}



Next request was
Request URL:https://sat61fusor.example.com/fusor/api/v21/deployments/2

{"deployment":{"id":2,"name":"deploy_12_23_a","description":null,"deploy_rhev":true,"deploy_cfme":true,"deploy_openstack":false,"rhev_engine_admin_password":"PW","rhev_database_name":"Default","rhev_cluster_name":"Default","rhev_storage_name":"my_storage","rhev_storage_type":"NFS","rhev_storage_address":"192.168.252.10","rhev_cpu_type":null,"rhev_share_path":"/var/lib/exports/vms","rhev_export_domain_name":"my_export","rhev_export_domain_address":"192.168.252.10","rhev_export_domain_path":"/var/lib/exports/exports","rhev_local_storage_path":null,"rhev_gluster_node_name":null,"rhev_gluster_node_address":null,"rhev_gluster_ssh_port":null,"rhev_gluster_root_password":null,"rhev_is_self_hosted":false,"cfme_install_loc":"RHEV","foreman_task_uuid":null,"upstream_consumer_uuid":"63170c3b-994d-4dd3-82ac-7cccf353206c","upstream_consumer_name":"jmatthews-rhci","rhev_root_password":"PW","cfme_root_password":"PW","cfme_admin_password":"PW","host_naming_scheme":"Freeform","custom_preprend_name":null,"enable_access_insights":true,"cfme_address":null,"rhev_engine_host_id":4,"openstack_undercloud_password":null,"openstack_undercloud_ip_addr":null,"openstack_undercloud_user":null,"openstack_undercloud_user_password":null,"openstack_overcloud_address":null,"openstack_overcloud_password":null,"openstack_overcloud_private_net":null,"openstack_overcloud_float_net":null,"openstack_overcloud_float_gateway":null,"openstack_undercloud_hostname":null,"openstack_overcloud_hostname":null,"cfme_hostname":null,"is_disconnected":false,"cdn_url":null,"manifest_file":null,"created_at":"2015-12-24T02:08:36Z","updated_at":"2015-12-24T02:09:41Z","organization_id":1,"lifecycle_environment_id":null,"discovered_host_id":4,"discovered_host_ids":[5],"subscription_ids":[],"introspection_task_ids":[]},"organizations":[{"id":1,"name":"Default Organization","title":"Default Organization","label":"Default_Organization","description":null,"created_at":"2015-12-22T16:50:06Z","updated_at":"2015-12-22T16:54:56Z"}],"lifecycle_environments":[],"discovered_hosts":[{"id":4,"name":"mac525400e3884f","type":"Host::Discovered","ip":"192.168.252.17","mac":"52:54:00:e3:88:4f","created_at":"2015-12-24T02:07:57Z","updated_at":"2015-12-24T02:07:57Z","cpus":1,"memory_human_size":"1.96 GB","disks_human_size":"10 GB","disk_count":1,"subnet_to_s":"default (192.168.252.0/24)","is_virtual":"true","environment_name":null,"hostgroup_name":null,"compute_resource_name":null,"domain_name":null,"is_managed":false,"is_discovered":true},{"id":5,"name":"mac525400967cd9","type":"Host::Discovered","ip":"192.168.252.18","mac":"52:54:00:96:7c:d9","created_at":"2015-12-24T02:08:27Z","updated_at":"2015-12-24T02:08:28Z","cpus":4,"memory_human_size":"7.8 GB","disks_human_size":"10 GB","disk_count":1,"subnet_to_s":"default (192.168.252.0/24)","is_virtual":"true","environment_name":null,"hostgroup_name":null,"compute_resource_name":null,"domain_name":null,"is_managed":false,"is_discovered":true}],"subscriptions":[],"introspection_tasks":[]}


Then:  

Request URL:https://sat61fusor.example.com/api/v21/foreman_tasks?search=parent_task_id+%3D+undefined

Response
{"foreman_tasks":[]}



Notice the 'undefined' in the task id.

isratrade commented:
@jwmatthews, I refactored and fixed the issue where `parent_task_id` should not be `undefined`. I kept the separate string attribute `foreman_task_uuid` rather than buying it in the belongsTo relationship as the `key`. I also squashed all commits to just one.

isratrade commented on fusor-ember-cli/app/serializers/foreman-task.js:
```diff
@@ -1,6 +1,7 @@
 
-export default DS.RESTSerializer.extend({
+export default DS.ActiveModelSerializer.extend({
+  isNewSerializerAPI: true,
```
I was getting errors if I didn't assign `isNewSerializerAPI: true`. Not sure what caused this change.

jwmatthews replied:
@isratrade I looked up the reason for isNewSerializerAPI: true:

http://emberjs.com/blog/2015/06/18/ember-data-1-13-released.html#toc_upgrade-guide

If you have customized your serializer, you should upgrade to Ember Data 1.13, check the upgrade guide to see if you need to make any changes, and then set a temporary flag on your Serializer: isNewSerializerAPI. This will opt you into the new serializer API. Once you are on the Ember Data 2.0 train, new Serializer API is the default one, and there is no need for a flag.

isratrade replied:
@jwmatthews. thanks John.

isratrade commented:
ping @jmontleon @jwmatthews if this is good to merge?

cfchase commented:
ACK
